src/compiler/transformers/classFields.ts

```diff
@@ -1095,6 +1148,75 @@ namespace ts {
+                    onEmitComputedPropertyName(hint, node as ComputedPropertyName, emitCallback);
+                    break;
+                case SyntaxKind.PropertyDeclaration:
+                    onEmitPropertyDeclaration(hint, node as PropertyDeclaration, emitCallback);
+                    break;
+                default:
+                    previousOnEmitNode(hint, node, emitCallback);
+            }
+
+        }
+
+        function onEmitPropertyDeclaration(hint: EmitHint, node: PropertyDeclaration, emitCallback: (hint: EmitHint, node: Node) => void) {
+            const savedStaticFieldInitializerContext = currentStaticFieldInitializerContext;
+            currentStaticFieldInitializerContext = staticFieldInitializerContextMap[getOriginalNodeId(node)];
+
+            previousOnEmitNode(hint, node, emitCallback);
+            currentStaticFieldInitializerContext = savedStaticFieldInitializerContext;
+        }
+
+        function onEmitFunctionLikeDeclaration(hint: EmitHint, node: FunctionDeclaration | FunctionExpression | ConstructorDeclaration, emitCallback: (hint: EmitHint, node: Node) => void) {
+            const savedStaticFieldInitializerContext = currentStaticFieldInitializerContext;
+            currentStaticFieldInitializerContext = undefined;
+
+            previousOnEmitNode(hint, node, emitCallback);
+            currentStaticFieldInitializerContext = savedStaticFieldInitializerContext;
+        }
+
+        function onEmitNamedMemberDeclaration(hint: EmitHint, node: AccessorDeclaration | MethodDeclaration, emitCallback: (hint: EmitHint, node: Node) => void) {
+            const savedStaticFieldInitializerContext = currentStaticFieldInitializerContext;
+            const savedCurrentComputedPropertyNameStaticFieldInitializerContext = currentComputedPropertyNameStaticFieldInitializerContext;
+            currentComputedPropertyNameStaticFieldInitializerContext = currentStaticFieldInitializerContext;
+            currentStaticFieldInitializerContext = undefined;
+
+            previousOnEmitNode(hint, node, emitCallback);
+            currentStaticFieldInitializerContext = savedStaticFieldInitializerContext;
+            currentComputedPropertyNameStaticFieldInitializerContext = savedCurrentComputedPropertyNameStaticFieldInitializerContext;
+        }
+
+        function onEmitComputedPropertyName(hint: EmitHint, node: ComputedPropertyName, emitCallback: (hint: EmitHint, node: Node) => void) {
+            const savedStaticFieldInitializerContext = currentStaticFieldInitializerContext;
+            currentStaticFieldInitializerContext = currentComputedPropertyNameStaticFieldInitializerContext;
+
+            previousOnEmitNode(hint, node, emitCallback);
+            currentStaticFieldInitializerContext = savedStaticFieldInitializerContext;
+        }
+
+        function onEmitThisOrSuperExpression(hint: EmitHint, node: ThisExpression | SuperExpression, emitCallback: (hint: EmitHint, node: Node) => void) {

```

Comment:
Is this one necessary? Shouldn't `currentStaticFieldInitializerContext` already be set as you descend? I'm not sure why you'd need to reset it.

Reply:
> Is this one necessary?

I guess so. I'd like to bind the visiting `currentStaticFieldInitializerContext` with the this or super expression to directly use when substitution.